bionic/aip/future.py

```diff
@@ -0,0 +1,96 @@
+import blocks
+import logging
+import time
+import googleapiclient.discovery
+from concurrent.futures import Future as _Future, TimeoutError, CancelledError
+from enum import Enum, auto
+
+
+class AIPError(Exception):
+    pass
+
+
+class State(Enum):
+    STATE_UNSPECIFIED = auto()
+    QUEUED = auto()
+    PREPARING = auto()
+    RUNNING = auto()
+    SUCCEEDED = auto()
+    FAILED = auto()
+    CANCELLING = auto()
+    CANCELLED = auto()
+
+
+EXECUTE = {State.STATE_UNSPECIFIED, State.QUEUED, State.PREPARING, State.RUNNING}
+CANCEL = {State.CANCELLING, State.CANCELLED}
+FAIL = {State.FAILED}
+
+
+class Future(_Future):
+    def __init__(self, project: str, job_id: str, output: str):
+        self.project = project
+        self.job_id = job_id
+        self.output = output
+        self.aip = googleapiclient.discovery.build("ml", "v1", cache_discovery=False)
+
+    def name(self):
+        return f"projects/{self.project}/jobs/{self.job_id}"
+
+    def cancel(self):
+        request = self.aip.projects().jobs().cancel(name=self.name())
+        request.execute()
+        return True
+
+    def query(self):
+        request = self.aip.projects().jobs().get(name=self.name())
+        response = request.execute()
+        return State[response["state"]], response.get("errorMessage", "")
+
+    def cancelled(self):
+        # TODO do we ever want to distinguish between canceling and cancelled on GCP?
+        # at the moment we just trust that GCP will succesfully cancel once requested
+        state, _ = self.query()
+        return state in CANCEL
+
+    def running(self):
+        state, _ = self.query()
+        return state in EXECUTE
+
+    def done(self):
+        state, _ = self.query()
+        return state is State.SUCCEEDED or state in CANCEL
+
+    def result(self, timeout: int = None):
+        # TODO support other serializers
+        exc = self.exception(timeout)
+        if exc is not None:
+            raise exc
+
+        try:
+            return blocks.unpickle(self.output)
+        except ValueError:
```

Comment:
This will raise a ValueError if the file does not exist at all, which is the primary thing i was trying to catch here. It doesn't seem bad though to do a generic catch in this case however since we will always reraise